ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java

```diff
@@ -0,0 +1,851 @@
+   * @param diskIOScheduler the {@link DiskIOScheduler} to schedule I/O.
+   * @param srcLog the {@link Log} to copy data from.
+   * @param scheduler the {@link ScheduledExecutorService} to use.
+   * @param recovery the {@link MessageStoreRecovery} to use to recover the index and log.
+   * @param time the {@link Time} instance to use.
+   * @param sessionId the sessionID of the store.
+   * @param incarnationId the incarnation ID of the store.
+   */
+  BlobStoreCompactor(String dataDir, String storeId, StoreKeyFactory storeKeyFactory, StoreConfig config,
+      StoreMetrics metrics, DiskIOScheduler diskIOScheduler, Log srcLog, ScheduledExecutorService scheduler,
+      MessageStoreRecovery recovery, Time time, UUID sessionId, UUID incarnationId) {
+    this.dataDir = new File(dataDir);
+    this.storeId = storeId;
+    this.storeKeyFactory = storeKeyFactory;
+    this.config = config;
+    this.metrics = metrics;
+    this.srcLog = srcLog;
+    this.diskIOScheduler = diskIOScheduler;
+    this.scheduler = scheduler;
+    this.recovery = recovery;
+    this.time = time;
+    this.sessionId = sessionId;
+    this.incarnationId = incarnationId;
+  }
+
+  /**
+   * If a compaction was in progress during a crash/shutdown, fixes the state so that the store is loaded correctly
+   * and compaction can resume smoothly on a call to {@link #resumeCompaction()}. Expected to be called before the
+   * {@link PersistentIndex} is created in the {@link BlobStore}.
+   * @throws IOException if the {@link CompactionLog} could not be created or if commit or cleanup failed.
+   * @throws StoreException if the commit failed.
+   */
+  void fixStateIfRequired() throws IOException, StoreException {
+    if (CompactionLog.isCompactionInProgress(dataDir.getAbsolutePath(), storeId)) {
+      compactionLog = new CompactionLog(dataDir.getAbsolutePath(), storeId, storeKeyFactory, time);
+      CompactionLog.Phase phase = compactionLog.getCompactionPhase();
+      switch (phase) {
+        case COMMIT:
+          commit(true);
+          compactionLog.markCleanupStart();
+        case CLEANUP:
+          cleanup(true);
+          compactionLog.markCycleComplete();
+          break;
+        case COPY:
+          recoveryStartToken = compactionLog.getSafeToken();
+        case PREPARE:
+        case DONE:
+          break;
+        default:
+          throw new IllegalStateException("Unrecognized compaction phase: " + phase);
+      }
+      if (compactionLog.getCompactionPhase().equals(CompactionLog.Phase.DONE)) {
+        endCompaction();
+      }
+    }
+  }
+
+  /**
+   * Starts the compactor and sets the {@link PersistentIndex} to copy data from.
+   * @param srcIndex the {@link PersistentIndex} to copy data from.
+   */
+  void start(PersistentIndex srcIndex) {
+    this.srcIndex = srcIndex;
+    if (compactionLog == null && srcIndex.hardDeleter != null && srcIndex.hardDeleter.isPaused()) {
+      srcIndex.hardDeleter.resume();
```

Comment:
Thanks for bringing this up. This will follow the same convention as the `StoreFindToken` instances associated with replication and will be reset appropriately. Currently, this will be reset to the beginning and hard delete works ok with that. It might redo the work however and we can improve upon that with the reset key changes.